### PR TITLE
XW-2223 | GlobalTitle - fix fetching namespaces

### PR DESCRIPTION
**Pull request by kvas-damian**
https://wikia-inc.atlassian.net/browse/XW-2223

Ping @Wikia/x-wing

### CODE REVIEW

**Comment by WikiaTech** [comment]
SonarQube analysis reported 151 issues
* CRITICAL 1 critical
* MAJOR 50 major
* MINOR 49 minor
* INFO 51 info

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [Language.php#L3639](https://github.com/Wikia/app/blob/bfce3221e1f7a823bd40b7dca9259091c61ed15a/languages/Language.php#L3639): Move arguments "$prefix" after arguments without default value [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1788)
1. MAJOR [GlobalTitle.php#L18](https://github.com/Wikia/app/blob/bfce3221e1f7a823bd40b7dca9259091c61ed15a/includes/wikia/GlobalTitle.php#L18): Class "GlobalTitle" has 35 methods, which is greater than 20 authorized. Split it into smaller classes. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1448)
1. MAJOR [GlobalTitle.php#L71](https://github.com/Wikia/app/blob/bfce3221e1f7a823bd40b7dca9259091c61ed15a/includes/wikia/GlobalTitle.php#L71): Define and throw a dedicated exception instead of using a generic one. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS112)
1. MAJOR [GlobalTitle.php#L100](https://github.com/Wikia/app/blob/bfce3221e1f7a823bd40b7dca9259091c61ed15a/includes/wikia/GlobalTitle.php#L100): Define and throw a dedicated exception instead of using a generic one. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS112)
1. MAJOR [GlobalTitle.php#L125](https://github.com/Wikia/app/blob/bfce3221e1f7a823bd40b7dca9259091c61ed15a/includes/wikia/GlobalTitle.php#L125): Define and throw a dedicated exception instead of using a generic one. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS112)
1. MAJOR [GlobalTitle.php#L686](https://github.com/Wikia/app/blob/bfce3221e1f7a823bd40b7dca9259091c61ed15a/includes/wikia/GlobalTitle.php#L686): Reduce the number of returns of this function 4, down to the maximum allowed 3. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1142)
1. MAJOR [GlobalTitle.php#L762](https://github.com/Wikia/app/blob/bfce3221e1f7a823bd40b7dca9259091c61ed15a/includes/wikia/GlobalTitle.php#L762): Remove the use of this variable variable "$$key". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1599)
1. MAJOR [Language.php](https://github.com/Wikia/app/blob/bfce3221e1f7a823bd40b7dca9259091c61ed15a/languages/Language.php): File "Language.php" has 4160 lines, which is greater than 1000 authorized. Split it into smaller files. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS104)
1. MAJOR [Language.php#L36](https://github.com/Wikia/app/blob/bfce3221e1f7a823bd40b7dca9259091c61ed15a/languages/Language.php#L36): Remove the unused function parameter "$variant". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1172)
1. MAJOR [Language.php#L43](https://github.com/Wikia/app/blob/bfce3221e1f7a823bd40b7dca9259091c61ed15a/languages/Language.php#L43): Remove the unused function parameters "$l, $n, $ignoreOtherCond". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1172)

File: includes/wikia/GlobalTitle.php [hunk 1]
```diff
@@ -822,10 +822,22 @@ private function loadNamespaceNames() {
 			return $this->mNamespaceNames;
 		}
 
+		// $wgMetaNamespace is calculated per wiki and cached in Language singleton per language,
+		// so we have to override it manually in this method
+		$metaNamespace = WikiFactory::getVarValueByName( "wgMetaNamespace", $this->mCityId );
+		$metaNamespaceTalk = WikiFactory::getVarValueByName( "wgMetaNamespaceTalk", $this->mCityId );
```
Comment (rogatty) [inline includes/wikia/GlobalTitle.php]: Let's stick to single quotes as seen below.

File: includes/wikia/GlobalTitle.php [hunk 2]
```diff
@@ -822,10 +822,22 @@ private function loadNamespaceNames() {
...
+
+		if ( $metaNamespace === false ) {
+			$wiki = WikiFactory::getWikiByID( $this->mCityId );
+			$metaNamespace = str_replace( ' ', '_', $wiki->city_title );
```
Comment (rogatty) [inline includes/wikia/GlobalTitle.php]: Please add a comment that this is duplicated from somewhere else and why.

File: includes/wikia/GlobalTitle.php [hunk 3]
```diff
@@ -822,10 +822,22 @@ private function loadNamespaceNames() {
...
 		}
 
+		// $wgMetaNamespace is calculated per wiki and cached in Language singleton per language,
+		// so we have to override it manually in this method
+		$metaNamespace = WikiFactory::getVarValueByName( 'wgMetaNamespace', $this->mCityId );
+		$metaNamespaceTalk = WikiFactory::getVarValueByName( 'wgMetaNamespaceTalk', $this->mCityId );
+
+		if ( $metaNamespace === false ) {
+			$wiki = WikiFactory::getWikiByID( $this->mCityId );
+			$metaNamespace = str_replace( ' ', '_', $wiki->city_title );
```
Comment (rogatty) [inline includes/wikia/GlobalTitle.php:833]: Please add a comment that this is duplicated from somewhere else and why.

**Comment by WikiaTech** [comment]
SonarQube analysis reported 151 issues
* CRITICAL 1 critical
* MAJOR 50 major
* MINOR 49 minor
* INFO 51 info

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [Language.php#L3639](https://github.com/Wikia/app/blob/a92c011aa12c3188e3998a2287ff1b6b8f1af388/languages/Language.php#L3639): Move arguments "$prefix" after arguments without default value [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1788)
1. MAJOR [GlobalTitle.php#L18](https://github.com/Wikia/app/blob/a92c011aa12c3188e3998a2287ff1b6b8f1af388/includes/wikia/GlobalTitle.php#L18): Class "GlobalTitle" has 35 methods, which is greater than 20 authorized. Split it into smaller classes. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1448)
1. MAJOR [GlobalTitle.php#L71](https://github.com/Wikia/app/blob/a92c011aa12c3188e3998a2287ff1b6b8f1af388/includes/wikia/GlobalTitle.php#L71): Define and throw a dedicated exception instead of using a generic one. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS112)
1. MAJOR [GlobalTitle.php#L100](https://github.com/Wikia/app/blob/a92c011aa12c3188e3998a2287ff1b6b8f1af388/includes/wikia/GlobalTitle.php#L100): Define and throw a dedicated exception instead of using a generic one. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS112)
1. MAJOR [GlobalTitle.php#L125](https://github.com/Wikia/app/blob/a92c011aa12c3188e3998a2287ff1b6b8f1af388/includes/wikia/GlobalTitle.php#L125): Define and throw a dedicated exception instead of using a generic one. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS112)
1. MAJOR [GlobalTitle.php#L686](https://github.com/Wikia/app/blob/a92c011aa12c3188e3998a2287ff1b6b8f1af388/includes/wikia/GlobalTitle.php#L686): Reduce the number of returns of this function 4, down to the maximum allowed 3. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1142)
1. MAJOR [GlobalTitle.php#L762](https://github.com/Wikia/app/blob/a92c011aa12c3188e3998a2287ff1b6b8f1af388/includes/wikia/GlobalTitle.php#L762): Remove the use of this variable variable "$$key". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1599)
1. MAJOR [Language.php](https://github.com/Wikia/app/blob/a92c011aa12c3188e3998a2287ff1b6b8f1af388/languages/Language.php): File "Language.php" has 4160 lines, which is greater than 1000 authorized. Split it into smaller files. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS104)
1. MAJOR [Language.php#L36](https://github.com/Wikia/app/blob/a92c011aa12c3188e3998a2287ff1b6b8f1af388/languages/Language.php#L36): Remove the unused function parameter "$variant". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1172)
1. MAJOR [Language.php#L43](https://github.com/Wikia/app/blob/a92c011aa12c3188e3998a2287ff1b6b8f1af388/languages/Language.php#L43): Remove the unused function parameters "$l, $n, $ignoreOtherCond". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1172)

**Review by hakubo — APPROVED** [review]

File: languages/Language.php [hunk 4]
```diff
@@ -137,18 +137,30 @@ class Language {
...
+	static function factory( $code, $cityId = null ) {
+		global $wgCityId;
+		if ( is_null( $cityId )) {
+			$cityId = $wgCityId;
+		}
+
+		if ( !isset( self::$mLangObjCache[$code][$cityId] ) ) {
 			if ( count( self::$mLangObjCache ) > 10 ) {
 				// Don't keep a billion objects around, that's stupid.
 				self::$mLangObjCache = array();
```
Comment (rogatty) [inline languages/Language.php]: Do we want to keep it or believe it's stupid to clean the cache like that? If we want to keep then let's make it `= []`

**Review by rogatty — APPROVED** [review]

**Comment by WikiaTech** [comment]
SonarQube analysis reported 151 issues
- CRITICAL 1 critical
- MAJOR 50 major
- MINOR 49 minor
- INFO 51 info

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:
1. CRITICAL [Language.php#L3639](https://github.com/Wikia/app/blob/75b4be5d768b947a1bcbdd4af3436fe62f6d1047/languages/Language.php#L3639): Move arguments "$prefix" after arguments without default value [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1788)
2. MAJOR [GlobalTitle.php#L18](https://github.com/Wikia/app/blob/75b4be5d768b947a1bcbdd4af3436fe62f6d1047/includes/wikia/GlobalTitle.php#L18): Class "GlobalTitle" has 35 methods, which is greater than 20 authorized. Split it into smaller classes. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1448)
3. MAJOR [GlobalTitle.php#L71](https://github.com/Wikia/app/blob/75b4be5d768b947a1bcbdd4af3436fe62f6d1047/includes/wikia/GlobalTitle.php#L71): Define and throw a dedicated exception instead of using a generic one. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS112)
4. MAJOR [GlobalTitle.php#L100](https://github.com/Wikia/app/blob/75b4be5d768b947a1bcbdd4af3436fe62f6d1047/includes/wikia/GlobalTitle.php#L100): Define and throw a dedicated exception instead of using a generic one. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS112)
5. MAJOR [GlobalTitle.php#L125](https://github.com/Wikia/app/blob/75b4be5d768b947a1bcbdd4af3436fe62f6d1047/includes/wikia/GlobalTitle.php#L125): Define and throw a dedicated exception instead of using a generic one. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS112)
6. MAJOR [GlobalTitle.php#L686](https://github.com/Wikia/app/blob/75b4be5d768b947a1bcbdd4af3436fe62f6d1047/includes/wikia/GlobalTitle.php#L686): Reduce the number of returns of this function 4, down to the maximum allowed 3. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1142)
7. MAJOR [GlobalTitle.php#L762](https://github.com/Wikia/app/blob/75b4be5d768b947a1bcbdd4af3436fe62f6d1047/includes/wikia/GlobalTitle.php#L762): Remove the use of this variable variable "$$key". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1599)
8. MAJOR [Language.php](https://github.com/Wikia/app/blob/75b4be5d768b947a1bcbdd4af3436fe62f6d1047/languages/Language.php): File "Language.php" has 4160 lines, which is greater than 1000 authorized. Split it into smaller files. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS104)
9. MAJOR [Language.php#L36](https://github.com/Wikia/app/blob/75b4be5d768b947a1bcbdd4af3436fe62f6d1047/languages/Language.php#L36): Remove the unused function parameter "$variant". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1172)
10. MAJOR [Language.php#L43](https://github.com/Wikia/app/blob/75b4be5d768b947a1bcbdd4af3436fe62f6d1047/languages/Language.php#L43): Remove the unused function parameters "$l, $n, $ignoreOtherCond". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1172)

**Pull request merged**